### PR TITLE
Still more `rustc_mir_dataflow` cleanups

### PR DESCRIPTION
**Pull request by nnethercote**
A sequel to #118203, #118230, #118638, and #131481.

I'm pleased with this PR. It cleans up a few things that have been bugging me for a while. It took quite some effort to find these improvements.

r? @cjgillot

### CODE REVIEW

**Comment by nnethercote** [comment]
The `await-call-tree` regression looks like it might be real. I ran Cachegrind locally but it wasn't informative. A bunch of functions got more instructions, a bunch got fewer, and there was no clear pattern. The regression is so small and only on a stress test so I don't think it's important.

@rustbot label: +perf-regression-triaged

File: compiler/rustc_mir_dataflow/src/impls/storage_liveness.rs [hunk 1]
```diff
@@ -283,15 +282,15 @@ impl<'tcx> Analysis<'tcx> for MaybeRequiresStorage<'_, 'tcx> {
...
 impl<'tcx> MaybeRequiresStorage<'_, 'tcx> {
     /// Kill locals that are fully moved and have not been borrowed.
-    fn check_for_move(&mut self, trans: &mut <Self as Analysis<'tcx>>::Domain, loc: Location) {
-        let body = self.borrowed_locals.body();
-        let mut visitor = MoveVisitor { trans, borrowed_locals: &mut self.borrowed_locals };
-        visitor.visit_location(body, loc);
+    fn check_for_move(&self, trans: &mut <Self as Analysis<'tcx>>::Domain, loc: Location) {
+        let borrowed_locals = self.borrowed_locals.clone().into_results_cursor(self.body);
+        let mut visitor = MoveVisitor { trans, borrowed_locals };
+        visitor.visit_location(self.body, loc);
```
Comment (cjgillot) [inline compiler/rustc_mir_dataflow/src/impls/storage_liveness.rs:283]: This changes the complexity of the analysis. The former code use a cursor, which remembers the current position. The latter code recreates a cursor, which implicitly needs to recompute results from bb0. Could we use a `RefCell` or something?

Comment (nnethercote) [reply compiler/rustc_mir_dataflow/src/impls/storage_liveness.rs]: Hmm, ok. I've reinstated the cursor within a `RefCell`.

**Comment by nnethercote** [comment]
I have added two new commits: one that reinstates the cursor within `MaybeRequireStorage`, and one that does minor cleanups.

**Comment by nnethercote** [comment]
@Jarcho talked on Zulip about needing analysis mutability for some Clippy analyses. So I will close this PR and redo the non-mutability bits in other PRs.

**Pull request closed without merging**